Fix #4764: Spotlight overlays have some UI issues

JishnuGoyal opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

fixes #4764 
Updated video demo:

https://user-images.githubusercontent.com/64526117/206852213-f7a89815-3992-4aed-b8d1-38d0c22e6d4f.mp4


Updated screenshots:

![WhatsApp Image 2022-11-26 at 5 07 26 PM (1)](https://user-images.githubusercontent.com/64526117/204086939-e5433525-0f28-4686-b599-4515e26cb173.jpeg)
![WhatsApp Image 2022-11-26 at 5 07 26 PM](https://user-images.githubusercontent.com/64526117/204086940-3fd0bb20-e15d-4a6f-a7b5-bd7237f0350c.jpeg)
![WhatsApp Image 2022-11-26 at 5 07 25 PM (1)](https://user-images.githubusercontent.com/64526117/204086942-64611338-c428-4e11-9646-65fdea536c61.jpeg)
![WhatsApp Image 2022-11-26 at 5 07 25 PM](https://user-images.githubusercontent.com/64526117/204086943-5f09b2a3-c4d5-4c8d-974b-bcf74adfd8b9.jpeg)

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

Portrait

![WhatsApp Image 2022-11-25 at 10 28 03 PM (1)](https://user-images.githubusercontent.com/64526117/204029760-d312e050-72b6-42d1-84bf-5cf96080eee7.jpeg)
![WhatsApp Image 2022-11-25 at 10 28 03 PM](https://user-images.githubusercontent.com/64526117/204029764-c3d74f21-5670-4e13-ac37-79ba3843eec6.jpeg)
![WhatsApp Image 2022-11-25 at 10 28 04 PM (1)](https://user-images.githubusercontent.com/64526117/204029767-943a2fb9-831b-4917-a9bd-24b1d866ba0e.jpeg)
![WhatsApp Image 2022-11-25 at 10 28 04 PM (2)](https://user-images.githubusercontent.com/64526117/204029770-8797e881-d388-49bc-a0e4-4fae0235f801.jpeg)
![WhatsApp Image 2022-11-25 at 10 28 04 PM](https://user-images.githubusercontent.com/64526117/204029773-367ee30f-75d0-472b-889d-ea791dae7f27.jpeg)
![WhatsApp Image 2022-11-25 at 10 28 05 PM](https://user-images.githubusercontent.com/64526117/204029774-b2591983-ac05-4607-9127-86423d1ac320.jpeg)
![WhatsApp Image 2022-11-25 at 10 31 51 PM](https://user-images.githubusercontent.com/64526117/204030109-73450d40-9317-4d27-872c-8daed98b5032.jpeg)

Land:

![WhatsApp Image 2022-11-25 at 10 35 15 PM](https://user-images.githubusercontent.com/64526117/204030662-c6c18c7b-80bd-48aa-904c-e50a435d883b.jpeg)
![WhatsApp Image 2022-11-25 at 10 35 16 PM (1)](https://user-images.githubusercontent.com/64526117/204030668-1c944b62-22f2-4af9-8956-58f88a94c2d7.jpeg)
![WhatsApp Image 2022-11-25 at 10 35 16 PM (2)](https://user-images.githubusercontent.com/64526117/204030672-90fcc908-1dd1-4a5c-8867-33598c43bb21.jpeg)
![WhatsApp Image 2022-11-25 at 10 35 16 PM](https://user-images.githubusercontent.com/64526117/204030673-39bba84b-f387-464e-b28d-7129297d93c4.jpeg)
![WhatsApp Image 2022-11-25 at 10 35 17 PM (1)](https://user-images.githubusercontent.com/64526117/204030678-972298fa-e223-4b8f-8eb9-297f05745fce.jpeg)
![WhatsApp Image 2022-11-25 at 10 35 17 PM](https://user-images.githubusercontent.com/64526117/204030681-5260b61b-9bce-46b5-83d7-ee4a9c98c650.jpeg)

mobile rtl:
![WhatsApp Image 2022-11-25 at 10 44 58 PM](https://user-images.githubusercontent.com/64526117/204031771-f4e2d2e1-e91d-4505-b9e9-78cf148092c6.jpeg)
![WhatsApp Image 2022-11-25 at 10 44 59 PM (1)](https://user-images.githubusercontent.com/64526117/204031775-452d7e7d-bcb0-4987-9431-3d38e07f5719.jpeg)
![WhatsApp Image 2022-11-25 at 10 44 59 PM (2)](https://user-images.githubusercontent.com/64526117/204031779-f4e75421-f8fe-4325-9f2b-67970889eb73.jpeg)
![WhatsApp Image 2022-11-25 at 10 44 59 PM](https://user-images.githubusercontent.com/64526117/204031782-852b142e-4b89-47ba-9de8-39ae9f949442.jpeg)
![WhatsApp Image 2022-11-25 at 10 45 00 PM (1)](https://user-images.githubusercontent.com/64526117/204031785-d5cfa361-815a-40cb-be42-0b52ac46e8d0.jpeg)
![WhatsApp Image 2022-11-25 at 10 45 00 PM (2)](https://user-images.githubusercontent.com/64526117/204031787-b03affa8-08f2-41a3-8e4a-a2d00d535f20.jpeg)
![WhatsApp Image 2022-11-25 at 10 45 00 PM](https://user-images.githubusercontent.com/64526117/204031790-ccb3fde6-ad6d-45d6-8d10-5d158023287f.jpeg)

Tablet:
![image](https://user-images.githubusercontent.com/64526117/204032348-aabac5c6-7264-4b9b-9be8-c35b83329446.png)
![image](https://user-images.githubusercontent.com/64526117/204032377-a668902f-e585-49ab-801a-bf475abdc9e0.png)
![image](https://user-images.githubusercontent.com/64526117/204032427-36cb7df7-442e-44b6-8b0b-a0231466a61a.png)
![image](https://user-images.githubusercontent.com/64526117/204032539-44953af2-ed7c-463e-8896-71d90c567a4c.png)
![image](https://user-images.githubusercontent.com/64526117/204032581-b2fa1837-ec57-495c-97ef-76027556124b.png)
![image](https://user-images.githubusercontent.com/64526117/204032637-ac5e0d05-d0de-4d2e-a4f5-2a0bf0f1c4d1.png)

Video (spotlight experience):

https://user-images.githubusercontent.com/64526117/204033854-12f2f0ca-fb92-4ddc-8589-02e5b1e38c46.mp4

Video (Talkback):

https://user-images.githubusercontent.com/64526117/204034102-f420b8ef-972e-4517-a55d-dcd3d5fb5919.mp4

JishnuGoyal committed:
first approach

JishnuGoyal committed:
before starting top down.

JishnuGoyal committed:
binding for spotlight onboarding

JishnuGoyal committed:
add dependency

JishnuGoyal committed:
onboarding spotlight activated

JishnuGoyal committed:
add platform parameter

JishnuGoyal committed:
topic spotlight activated

JishnuGoyal committed:
refactored and ready to document

JishnuGoyal committed:
refactored and ready to document

JishnuGoyal committed:
refactored and ready to document

JishnuGoyal committed:
multiple overlay res added and partially working

JishnuGoyal committed:
using single overlay. Calculate screen size method

JishnuGoyal committed:
spotlights in top position are working

JishnuGoyal committed:
use absolute screen position instead of view.x which was relative to parent distance

JishnuGoyal committed:
before adding rtl

JishnuGoyal committed:
rtl support added but not working

JishnuGoyal committed:
support for hint text workig

JishnuGoyal committed:
updated proto structure

JishnuGoyal committed:
updated model and domain layers

JishnuGoyal committed:
update onboarding and topic packages for new proto

JishnuGoyal committed:
spotlight fragment introduced. Retrieving from spotFrag so not working

JishnuGoyal committed:
rearrange

JishnuGoyal committed:
commit because i need to see another commit

JishnuGoyal committed:
commit before checkout

JishnuGoyal committed:
store position parameters as fields

JishnuGoyal committed:
all working. Partial impl for rtl compat

JishnuGoyal committed:
rtl functionality added

JishnuGoyal committed:
support for suspending spotlights when talkback is being used

JishnuGoyal committed:
use indexed list parser

JishnuGoyal committed:
initial commit - to be reverted

JishnuGoyal committed:
setup

JishnuGoyal committed:
setup spotlight fragment

JishnuGoyal committed:
setup - not to be pushed

JishnuGoyal committed:
add lesson tab spotlight and content desc

JishnuGoyal committed:
add test

JishnuGoyal committed:
try to get reference to the toolbar through findviewbyid

JishnuGoyal committed:
re-structure the code

JishnuGoyal committed:
rename overlay layout resources

JishnuGoyal committed:
lint

JishnuGoyal committed:
remove unused resources

JishnuGoyal committed:
toolbar can now be spotlit. todo: Add extra logic for including some spotlight targets

JishnuGoyal committed:
do not push

JishnuGoyal committed:
spotlight languages icon

JishnuGoyal committed:
Merge branch 'develop' into introduce_spotlight_fragment

JishnuGoyal committed:
spotlight setup complete for base prs

JishnuGoyal committed:
add index to story, spotlight chapter using custom view

JishnuGoyal committed:
remove unused resources

JishnuGoyal committed:
Merge branch 'develop' into promoted_story_spotlight_custom_view

JishnuGoyal committed:
setup

JishnuGoyal committed:
add spotlight and tests

JishnuGoyal committed:
ben's suggestion

JishnuGoyal committed:
revision and lesson tab spotlights working. Tests to be added.

JishnuGoyal committed:
spotlights working. tests to be added.

JishnuGoyal committed:
Merge branch 'lesson_screen_spotlights' into add_all_spotlights

JishnuGoyal committed:
Merge branch 'audio_fragment_spotlights' into add_all_spotlights

# Conflicts:
#	app/src/main/java/org/oppia/android/app/topic/TopicFragment.kt

JishnuGoyal committed:
no changes added

JishnuGoyal committed:
Merge branch 'use_custom_view_to_spotlight_first_chapter' into add_all_spotlights

# Conflicts:
#	app/src/main/java/org/oppia/android/app/view/ViewComponentImpl.kt

JishnuGoyal committed:
all spotlights work.

JishnuGoyal committed:
add target store impl

JishnuGoyal committed:
custom queue implementation done

JishnuGoyal committed:
small fixes

JishnuGoyal committed:
Merge branch 'spotlight_custom_queue' into add_all_spotlights_using_custom_queue

# Conflicts:
#	app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt
#	app/src/main/java/org/oppia/android/app/topic/TopicFragment.kt
#	app/src/main/res/layout/overlay.xml

JishnuGoyal committed:
add all spotlights, properly initialize profileid in spotlight fragment, inject fields into spotlights fragment

JishnuGoyal committed:
update spotlight fragment and refactor function name

JishnuGoyal committed:
lint

JishnuGoyal committed:
manage audio that was playing beneath the spotlight fragment

JishnuGoyal committed:
commit before checkout

JishnuGoyal committed:
slight adjustments for overlay design

JishnuGoyal committed:
finalise overlay alignments

JishnuGoyal committed:
adjustments so that app temporarily works

JishnuGoyal committed:
restrict doOnPreDraw calls to spotlight fragment

JishnuGoyal committed:
improve placement of overlay elements

JishnuGoyal committed:
top left position in rtl not working

JishnuGoyal committed:
fully configure rtl compatibility

JishnuGoyal committed:
fully configure rtl compatibility + add onboarding spotlights

JishnuGoyal committed:
partial: add home activity and topic activity tests for spotlight.

JishnuGoyal committed:
spotlight voice-over icon after 3 logins

JishnuGoyal committed:
add tests to check spotlight is shown

JishnuGoyal committed:
problem with tests

JishnuGoyal committed:
add create functions for the spotlight fragments

JishnuGoyal committed:
commit review suggestions

JishnuGoyal committed:
dont use synthetic

JishnuGoyal committed:
make sure voiceover icon is visible as well

JishnuGoyal committed:
add forced announcement for hints

JishnuGoyal committed:
move test to another file

JishnuGoyal committed:
add all tests for individual spotlights

JishnuGoyal committed:
overlay design changes

JishnuGoyal committed:
some tests in exploration activity tests are failing

JishnuGoyal committed:
adjust tests for homeactivitytest

JishnuGoyal committed:
merge with develop

JishnuGoyal committed:
re adjust test

JishnuGoyal committed:
setup tests for spotlight fragment

JishnuGoyal committed:
add kdocs

JishnuGoyal committed:
use correct test configuration

JishnuGoyal committed:
suggetive fixes

JishnuGoyal committed:
add bazel config

add bazel config

JishnuGoyal committed:
adjust affected tests

JishnuGoyal committed:
Merge remote-tracking branch 'origin/add_all_spotlights_using_custom_queue' into add_all_spotlights_using_custom_queue

JishnuGoyal committed:
merge with develop

JishnuGoyal committed:
adjust tests

JishnuGoyal committed:
kdocs

JishnuGoyal committed:
fix bug

JishnuGoyal committed:
small corrections

JishnuGoyal committed:
adjust test

JishnuGoyal committed:
use resource handler instead

JishnuGoyal committed:
lint

JishnuGoyal committed:
use spannable to highlight text

JishnuGoyal committed:
suggestions

JishnuGoyal committed:
lint

BenHenning committed:
Fix broken test.

To ensure doOnPreDraw actually runs, we need to request that the view
gets re-drawn.

BenHenning committed:
Fix variety of issues.

JishnuGoyal committed:
Merge pull request #4 from BenHenning/small-fixes

Small fixes

madhurgera2 committed:
lint

madhurgera2 committed:
Merge remote-tracking branch 'origin/add_all_spotlights_using_custom_queue' into add_all_spotlights_using_custom_queue

# Conflicts:
#	app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt

madhurgera2 committed:
fix build failure

madhurgera2 committed:
fix build failure

JishnuGoyal committed:
Revert "Fix #4746: Update app to target SDK 31 (#4747)"

This reverts commit 7cb38ffa

Commit:
add hint background design

Commit:
Merge branch 'develop' into add_all_spotlights_using_custom_queue

# Conflicts:
#	app/BUILD.bazel
#	app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt
#	app/src/main/java/org/oppia/android/app/view/ViewComponentImpl.kt
#	app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
#	scripts/assets/test_file_exemptions.textproto

Commit:
merge with develop

JishnuGoyal committed:
suggestive comments

JishnuGoyal committed:
suggestive comments

JishnuGoyal committed:
general nits

JishnuGoyal committed:
general nits

JishnuGoyal committed:
general nits

JishnuGoyal committed:
general nits

JishnuGoyal committed:
add requested changes

JishnuGoyal committed:
more fixes

JishnuGoyal committed:
more fixes

JishnuGoyal committed:
more fixes

BenHenning committed:
Fix broken ExplorationActivityTest tests.

BenHenning committed:
Revert "Revert "Fix #4746: Update app to target SDK 31 (#4747)""

This reverts commit 912c357300f408f73a2aa4d6e94727c6605d9cdb.

JishnuGoyal committed:
more fixes

BenHenning committed:
Simplify margin computations a bit for overlays.

JishnuGoyal committed:
another fix

JishnuGoyal committed:
another fix

BenHenning committed:
Merge branch 'add_all_spotlights_using_custom_queue' of https://github.com/JishnuGoyal/oppia-android into some-additional-small-fixes

Conflicts:
	app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt

JishnuGoyal committed:
fix constraints, add tests

BenHenning committed:
Merge branch 'add_all_spotlights_using_custom_queue' of https://github.com/JishnuGoyal/oppia-android into some-additional-small-fixes

BenHenning committed:
Miscellaneous follow-up static check & test fixes.

JishnuGoyal committed:
some additional small fixes

Some additional small fixes

BenHenning committed:
Merge branch 'develop' into add_all_spotlights_using_custom_queue

JishnuGoyal committed:
fixes

JishnuGoyal committed:
final suggestions

JishnuGoyal committed:
final suggestions

JishnuGoyal committed:
suggested fixes

JishnuGoyal committed:
remove onboarding spotlight

JishnuGoyal committed:
Merge branch 'develop' into add_all_spotlights_using_custom_queue

# Conflicts:
#	app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt
#	app/src/main/res/layout/bottom_left_overlay.xml
#	app/src/main/res/layout/bottom_right_overlay.xml
#	app/src/main/res/layout/onboarding_activity.xml
#	app/src/main/res/layout/top_left_overlay.xml
#	app/src/main/res/layout/top_right_overlay.xml
#	app/src/main/res/values/strings.xml
#	app/src/sharedTest/java/org/oppia/android/app/onboarding/OnboardingFragmentTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt

JishnuGoyal committed:
fix conflicts introduced with onboarding screen spotlight after merge with develop

JishnuGoyal committed:
fix conflicts introduced with onboarding screen spotlight after merge with develop

JishnuGoyal committed:
finalise done button

JishnuGoyal committed:
cross button

JishnuGoyal committed:
finalise cross button design

JishnuGoyal commented:
Update: All done.

> tasks left: Add screenshots and videos demonstrating the behaviour and UI of the spotlights introduced.

JishnuGoyal committed:
use spotlight manager instead

BenHenning commented on app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt:
```diff
@@ -137,7 +137,7 @@ class SpotlightFragment : InjectableFragment(), SpotlightNavigationListener, Spo
 
         override fun onEnded() {
-          targetList.removeFirst()
+          if (targetList.isNotEmpty()) targetList.removeFirst()
```
Is there a test that produces the crash which we can use this to verify as passing?

JishnuGoyal replied:
![image](https://user-images.githubusercontent.com/64526117/204085913-6b5e2969-d915-4770-97df-6fbee7178af6.png)

This tests mimics the behavior when the close spotlight button is tapped twice rapidly in a row. The test fails because the close button is not available, not because it is out of elements to pop. This essentially means that the functionality works as expected, but at the moment, more information is required how the crash reported in the email occurred. It might be possible that the crash was due to some unforeseen circumstance/state that was not caught by the error log such as too many apps being run parallelly on the device. In essence, the spotlight library itself has been designed in a way that this error should not occur. A test case cannot be written with the information we have today.

BenHenning replied:
Hmm okay. We could try to find a situation that would cause the specific crash in question, but I'm actually not sure how it could occur, either, and I think the fix is a reasonable mitigation regardless. I think it's okay that we don't add a test for this.

BenHenning commented on model/src/main/proto/spotlight.proto:
```diff
@@ -10,29 +10,26 @@ option java_multiple_files = true;
-
     // Corresponds to the topic fragment's lessons tab.
-    SpotlightViewState topic_lesson_tab = 2;
+    SpotlightViewState topic_lesson_tab = 1;
```
Keep existing indexes and mark the old onboarding one as reserved with a comment.

JishnuGoyal replied:
Done - Reserved in spotlightstatedatabase proto. The one mentioned here is a oneOf, from where it was removed. PTAL

rt4914 replied:
I have checked this and this change does look good.

BenHenning replied:
I think part of it is missing, left one follow-up.

BenHenning reviewed (changes requested):
Thanks @JishnuGoyal! I think it looks much better! I have a few comments, but besides that the alignment of text looks a bit off, ditto for the new x button: can we keep the top of the x button aligned with the text? Also, the text top and bottom margins relative to the background should always be the same (sometimes it seems like one is larger than the other in your pictures).

JishnuGoyal committed:
commit before checkout

rt4914 reviewed (approved):
Awesome work @JishnuGoyal Thanks.

JishnuGoyal committed:
requested changes

JishnuGoyal commented:
Thanks! @rt4914 and @BenHenning. FYI i have updated the UI as requested and addressed the review comments. This PR is getting a bit time sensitive so i have updated everything to the best of my knowledge. PTAL.

rt4914 commented:
https://github.com/oppia/oppia-android/pull/4763#discussion_r1032774542
-> Don't have much context about it so haven't been able to help to resolve the comment.

https://github.com/oppia/oppia-android/pull/4763#discussion_r1032689188
-> LGTM from my side.

BenHenning commented on model/src/main/proto/spotlight.proto:
```diff
@@ -10,9 +10,6 @@ option java_multiple_files = true;
   // Determines the UI element being spotlit.
   oneof feature {
-    // Corresponds to the onboarding screen's next button.
-    SpotlightViewState onboarding_next_button = 1;
```
Need reserved for this proto, too (it would go in Spotlight and not in the oneof).

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -583,7 +583,6 @@
-  <string name="onboarding_next_button_spotlight_hint">Next</string>
   <string name="state_fragment_hint_bar_forced_announcement_text">Go to the bottom of the screen for a hint.</string>
-  <string name="mark_spotlight_seen_button_text">Done</string>
+  <string name="mark_spotlight_seen_button_text">Close</string>
```
Is this still the seen button text or is it actually the content description for the new icon? Suggest either renaming or removing, based on that.

JishnuGoyal replied:
removing since spotlights themselves are suspended during talkback

BenHenning reviewed (commented):
Thanks @JishnuGoyal. Just two follow-ups--PTAL. Otherwise, the PR LGTM.

JishnuGoyal committed:
suggested changes

BenHenning reviewed (approved):
Thanks, LGTM!
